src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1361,15 +1434,31 @@ public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
                     // Just compute the checksum from the source text passed in.
                     originalSourceTextChecksum: null,
                     generationDateTime);
-                documentStates.Add(newGeneratedState);
+
+                if (syntaxNode != null)
+                {
+                    newGeneratedState = newGeneratedState.WithSyntaxRoot(syntaxNode);
+                }
+
+                documentStates.Add(newGeneratedState.Id, newGeneratedState);
+            }
+        }
+
+        if (FrozenSourceGeneratedDocumentStates is not null)
+        {
+            // We also carry forward any previously frozen source generated documents that were not part of this freeze. This ensures multiple
+            // freezes are additive, and everything will be unfrozen correctly, without us needing to have a chain of compilation trackers
+            foreach (var (id, state) in FrozenSourceGeneratedDocumentStates.States)
+            {
+                documentStates.TryAdd(id, state);
```

Comment:
wait.  is this correct?  if it's tryadd, coudl that mean a previous freeze already added it, and now the later one won't?  or is that simply not a problem here?

Reply:
`documentStates` is newed up in this method, and the first thing that happens is we add the document states modified in this round. The TryAdd calls come after, so will fill in any missing states from a previous round, without overriding.